merlin-server/sql/merlin/functions/hasura/hasura_functions.sql

```diff
@@ -27,3 +27,58 @@ begin
 	  order by p.name, ps.start_offset desc;
 end
 $$;
+
+create function hasura_functions.restore_activity_directive_from_changelog(
+  _plan_id integer,
+  _activity_directive_id integer,
+  _revision integer,
+  hasura_session json)
+  returns setof activity_directive
+  language plpgsql
+  volatile
+as $$
+declare
+  _function_permission metadata.permission;
+begin
+  _function_permission :=
+      metadata.get_function_permissions('restore_activity_directive_from_changelog', hasura_session);
+  if not _function_permission = 'NO_CHECK' then
+    call metadata.check_general_permissions(
+        'restore_activity_directive_from_changelog',
+        _function_permission, _plan_id,
+        (hasura_session ->> 'x-hasura-user-id')
+      );
+  end if;
+
+  if not exists(select revision
+                from public.activity_directive_changelog
+                where (plan_id, activity_directive_id, revision) =
+                      (_plan_id, _activity_directive_id, _revision))
+  then
+    raise exception 'Changelog Revision % does not exist for Plan % and Activity %', _revision, _plan_id, _activity_directive_id;
+  end if;
```

Comment:
You may want to check that the plan and activity directive exist so this method can throw a more understandable error if either don't.

Reply:
I figured this was logically equivalent to just checking the `activity_directive_changelog` table, due to foreign key constraints.

Reply:
For Plan, it is not, as proven by the test `shouldHaveNoChangelogsForEmptyPlan`.

For Activity Directive, while the test `shouldCreateChangelogForInsertedActDir` supports the idea that there should be at least one row in the changelog per revision, no rows in the table does not imply that the directive does not exist, as the table could've been deleted from by something other than the insert trigger.

Additionally, the given error messages do not communicate to the user that the plan or activity directive do not exist.

Reply:
Change has been made. By "logically equivalent" I meant that my original checks would still prevent the user from calling this function with invalid params.